### PR TITLE
update to allow non-determinstic ordering of socket warnings

### PR DESCRIPTION
**Pull request by eteq**
modification to astropy/astropy#10754 to allow the warnings to be in any order

### CODE REVIEW

File: astropy/utils/tests/test_data.py [hunk 1]
```diff
@@ -1055,16 +1055,22 @@ def osraiser(dirnm, linkto, pkgname=None):
...
-        idx_w1 = 0
-        idx_w2 = 3
-    else:
-        raise AssertionError(f'Expected 2 or 4 warnings, got {n_warns}')
-    assert 'Remote data cache could not be accessed' in str(warning_lines[idx_w1].message)
-    assert 'temporary' in str(warning_lines[idx_w2].message)
+
+    assert n_warns in (2, 4), f'Expected 2 or 4 warnings, got {n_warns}'
+
+    warning_msgs = [wl.message for wl in warning_lines]
```
Comment (pllim) [inline astropy/utils/tests/test_data.py]: Not sure if you need to explicitly cast it to str or not. I usually do.
```suggestion
    warning_msgs = [str(wl.message) for wl in warning_lines]
```

File: astropy/utils/tests/test_data.py [hunk 2]
```diff
@@ -1055,16 +1055,22 @@ def osraiser(dirnm, linkto, pkgname=None):
...
+
+    assert 'Remote data cache could not be accessed' in warning_msgs
```
Comment (pllim) [inline astropy/utils/tests/test_data.py]: I am not convinced partial match would work like this.

Comment (eteq) [reply astropy/utils/tests/test_data.py]: You're absolutely right, I wasn't paying attention and was thinking those are the full error messages - fix incoming

**Comment by eteq** [comment]
alright, it's now passing for me locally, @pllim. I even squashed some commits for you :wink:

File: astropy/utils/tests/test_data.py [hunk 3]
```diff
@@ -1055,16 +1055,20 @@ def osraiser(dirnm, linkto, pkgname=None):
...
-    elif n_warns == 4:  # Two mysterious unclosed socket warnings sometimes
-        idx_w1 = 0
-        idx_w2 = 3
-    else:
-        raise AssertionError(f'Expected 2 or 4 warnings, got {n_warns}')
-    assert 'Remote data cache could not be accessed' in str(warning_lines[idx_w1].message)
-    assert 'temporary' in str(warning_lines[idx_w2].message)
+
+    assert n_warns in (2, 4), f'Expected 2 or 4 warnings, got {n_warns}'
+
```
Comment (pllim) [inline astropy/utils/tests/test_data.py:1060]:
```suggestion
```

File: astropy/utils/tests/test_data.py [hunk 4]
```diff
@@ -1055,16 +1055,20 @@ def osraiser(dirnm, linkto, pkgname=None):
...
+
+    partial_warn_msgs = ['remote data cache could not be accessed' ,'temporary file']
```
Comment (pllim) [inline astropy/utils/tests/test_data.py]:
```suggestion
    partial_warn_msgs = ['remote data cache could not be accessed', 'temporary file']
```

File: astropy/utils/tests/test_data.py [hunk 5]
```diff
@@ -1055,16 +1055,20 @@ def osraiser(dirnm, linkto, pkgname=None):
...
+    if n_warns == 4:
+        partial_warn_msgs.extend(['socket', 'socket'])
+
+    for wl in warning_lines:
```
Comment (pllim) [inline astropy/utils/tests/test_data.py:1065]:
```suggestion
    for wl in warning_lines:
        cur_w = str(wl).lower()
```

File: astropy/utils/tests/test_data.py [hunk 6]
```diff
@@ -1055,16 +1055,20 @@ def osraiser(dirnm, linkto, pkgname=None):
...
+        for i, partial_msg in enumerate(partial_warn_msgs):
+            if partial_msg in str(wl).lower():
```
Comment (pllim) [inline astropy/utils/tests/test_data.py]:
```suggestion
            if partial_msg in cur_w:
```

File: astropy/utils/tests/test_data.py [hunk 7]
```diff
@@ -1055,16 +1055,20 @@ def osraiser(dirnm, linkto, pkgname=None):
...
+                del partial_warn_msgs[i]
+                break
+    assert len(partial_warn_msgs) == 0, 'Did not get some expected warnings:' + str(partial_warn_msgs)
```
Comment (pllim) [inline astropy/utils/tests/test_data.py]:
```suggestion
    assert len(partial_warn_msgs) == 0, f'Got some unexpected warnings: {partial_warn_msgs}'
```

**Review by pllim — APPROVED** [review]
Thanks! I'll just get these nitpicks in now and merge.

**Comment by pllim** [comment]
Thanks! That's a pretty smart way to check it. 😉

**Pull request merged**